data/mp/multiplay/skirmish/bonecrusher/functions.js

```diff
@@ -0,0 +1,1215 @@
+//к "живым" союзникам, полезно для отказа от захвата ресурсов союзника
+function filterNearAlly(obj){
+	for ( var p = 0; p < maxPlayers; ++p ){
+		if ( p == me ) continue; //Выкидываем себя
+		if ( !allianceExistsBetween(me,p) ) continue; //Выкидываем вражеские
+//		if ( playerLoose(p) ) continue; //Пропускаем проигравших
+		if ( playerSpectator(p) ) continue;
+		if ( distBetweenTwoPoints_p(base.x,base.y,startPositions[p].x,startPositions[p].y) < base_range ) continue; //Если союзник внутри радиуса нашей базы, вышки забираем
+		obj = obj.filter(function(e){if(distBetweenTwoPoints_p(e.x,e.y,startPositions[p].x,startPositions[p].y) < base_range )return false; return true;})
+	}
+	return obj;
+}
+
+//Функция отфильтровывает объекты внутри радиуса базы
+function filterNearBase(obj){
+	return obj.filter(function(e){
+		if( distBetweenTwoPoints_p(base.x,base.y, e.x, e.y) > base_range) return true; return false;
+	});
+}
+
+//функция отфильтровывает недостежимые объекты
+function filterInaccessible(obj){
+	return obj.filter(function(e){
+		//Если попыток постройки больше 3, отфильтровываем их
+		if(getInfoNear(e.x,e.y,'buildRig',0,300000,false,false,false).value > 3)return false;return true;
+	});
+}
+
+//TODO need to fix logic, sometimes it's buggy
+//it's triggered in middle of battlefield,
+//and send help to random location.. don't known why..
+function getEnemyNearAlly(){
+//	return []; // <-- disable this funtion
+	var targ = [];
+	var enemy = [];
+	for ( var e = 0; e < maxPlayers; ++e ) {
+		if ( allianceExistsBetween(me,e) ) continue;
+		targ = targ.concat(enumDroid(e, DROID_WEAPON, me));
+	}
+	if(scavengers == true) {
+		targ = targ.concat(enumDroid(scavengerPlayer, DROID_WEAPON, me));
+	}
+	
+	for ( var p = 0; p < maxPlayers; ++p ) {
+		if ( p == me ) continue;
+		if ( !allianceExistsBetween(me,p) ) continue;
+//		if ( playerLoose(p) ) continue; //Пропускаем проигравших
+		if ( playerSpectator(p) ) continue;
+//		enemy = enemy.concat(targ.filter(function(e){if(distBetweenTwoPoints_p(e.x,e.y,startPositions[p].x,startPositions[p].y) < (base_range/2) ){
+		enemy = enemy.concat(targ.filter(function(e){if(distBetweenTwoPoints_p(e.x,e.y,startPositions[p].x,startPositions[p].y) < base_range ){
+//			debugMsg("TRUE name="+e.name+"; id="+e.id+"; dist="+distBetweenTwoPoints_p(e.x,e.y,startPositions[p].x,startPositions[p].y)+"<"+(base_range/2)+"; player="+p,'temp');
+			return true;
+		}
+//		debugMsg("FALSE name="+e.name+"; id="+e.id+"; dist="+distBetweenTwoPoints_p(e.x,e.y,startPositions[p].x,startPositions[p].y)+"<"+(base_range/2)+"; player="+p, 'temp');
+		return false;}));
+	}
+	
+	return enemy;
+}
+
+//Функция возвращяет вышки, о которых в данный момент не известно ничего
+//Просто сравниваем два массива объектов и фильтруем в третий
+function getUnknownResources(){
+	var notSee = new Array();
+	var iSee = getSeeResources();
+	if ( iSee.length == 0 ) return allResources;
+	notSee = allResources.filter(function (value) {
```

Comment:
Can put `var` here and remove the first line in this function.

Reply:
Ok, fairly